nixos/modules/system/activation/switch-to-configuration.py

```diff
@@ -0,0 +1,611 @@
+
+    units_to_stop = transitions.units_to_stop()
+    units_to_start = transitions.units_to_start()
+    units_to_restart = transitions.units_to_restart()
+    units_to_skip = transitions.units_to_skip()
+
+    # Show dry-run actions.
+    if action == "dry-activate":
+        print_err(transitions.summary())
+        if restart_systemd:
+            print_err("would restart systemd")
+        sys.exit()
+
+    syslog(LOG_NOTICE, "switching to system configuration {out}".format(out=out))
+
+    if len(transitions.units_to_stop()) > 0:
+        if len(transitions.filtered_units_to_stop()) > 0:
+            print_err("stopping the following units: ",
+                    ', '.join(transitions.filtered_units_to_stop()))
+        # Use current version of systemctl binary before daemon is reexeced.
+        subprocess.check_call(["/run/current-system/sw/bin/systemctl", "stop", "--"] +
+                            sorted(transitions.units_to_stop())) # FIXME: ignore errors?
+
+    if len(transitions.units_to_skip()) > 0:
+        print_err("NOT restarting the following changed units: ",
+                ', '.join(sorted(transitions.units_to_skip())))
+
+    # Activate the new configuration (i.e., update /etc, make accounts,
+    # and so on).
+    print_err("activating the configuration...")
+    res = 0
+    if subprocess.call([out / "activate", out]) != 0:
+        res = 2
+
+    # Restart systemd if necessary.
+    if restart_systemd:
+        print_err("restarting systemd...")
+        if subprocess.call(["@systemd@/bin/systemctl", "daemon-reexec"]) != 0:
+            res = 2
+
+    # Forget about previously failed services.
+    subprocess.call(["@systemd@/bin/systemctl", "reset-failed"])
+
+    # Make systemd reload its units.
+    if subprocess.call(["@systemd@/bin/systemctl", "daemon-reload"]) != 0:
+        res = 3
+
+    reload_user_units()
+
+    # Set the new tmpfiles
+    print_err("setting up tmpfiles")
+    if subprocess.call(["@systemd@/bin/systemd-tmpfiles", "--create", "--remove",
+                        "--exclude-prefix=/dev"]) != 0:
+        res = 3
+
+    # Reload units that need it. This includes remounting changed mount
+    # units.
+    if len(transitions.units_to_reload()) > 0:
+        units = sorted(transitions.units_to_reload())
+        print_err("reloading the following units: ", ", ".join(units))
+        if subprocess.call(["@systemd@/bin/systemctl", "reload", "--"] +
+                           cast(List[str], units)) != 0:
+            res = 4
+        os.unlink(RELOAD_UNIT_FILE)
+
+    # Restart changed services (those that have to be restarted rather
+    # than stopped and started).
+    if len(transitions.units_to_restart()) > 0:
+        units = sorted(transitions.units_to_restart())
+        print_err("restarting the following units: ", ", ".join(units))
+        if subprocess.call(["@systemd@/bin/systemctl", "restart", "--"] +
+                           cast(List[str], units)) != 0:
```

Comment:
I think `["@systemd@/bin/systemctl", "restart", "--"] + list(units)` is a bit easier to read.